### ISSUE TITLE
Russian-doll-caching discussion

### ISSUE THREAD

**Issue by ekampp** [jsonapi-serializer/jsonapi-serializer#111]
This is a prompt for a discussion. We're using this library heavily, and we're beginning to see the need for a form of Russian-doll-caching of the different parts of the JSON within the same serializer. E.i.

```ruby
class ImportSerializer
  include FastJsonapi::ObjectSerializer

  cache_options \
    enabled: true,
    cache_length: 1.week

  attributes :attr1, :attr2

  has_one :target
  has_one :user
end
```

In the above case, it would be great if the `target` and `user`, `attr1`, and `attr2` could be individually cached, so that when let's say `attr2` changes, we don't have to go to the database to fetch the relations, but can read the new value directly from the object that's being serialized. 

Is this something that somehow is already implemented? In that case, please direct me to where I can find this. Otherwise, I would love to have a discussion about how this could be useful to the community, and how we could implement this. 

Thanks for reading, and hope to talk. 

Best, 
Emil

**Comment by stas** [comment]
@ekampp I'm not sure I follow here, maybe you can share some more details please. Particularly:
> so that when let's say attr2 changes, we don't have to go to the database to fetch the relations, but can read the new value directly from the object that's being serialized

The way caching works (on Rails at least) is that there's a key which includes the timestamp (say, `updated_at`, along with the cache store expiration/TTL). This means that the cache store is not evaluating each attribute or relationships fully, it's evaluating the `cache_key` (see [1], [2]). You can obviously change the `cache_key` implementation in the userland, but to make changes at the library level, more context is needed (at least for me).

Another detail here, is that fragment caching might make sense when serializing a collection or a single record has an `?include=`  param to sideload some records. But even here, the recent work on #90, showed that it still requires the userland to be aware what is being requested, since any of those changes will practically recache the response.

In a nutshell, caching works very transparently in `JSONAPI/Serializer` and it's pretty much described by the [ActiveSupport::Cache::Store](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html). The relevant conversation was at #50, please take a look if you have some time.

Let me know if I can help clarify anything :bow: 

[1]: https://devdocs.io/rails~6.0/activerecord/integration#method-i-cache_key
[2]: https://devdocs.io/rails~6.0/activerecord/relation#method-i-cache_key

**Comment by ekampp** [comment]
Calling `ImportSerializer.new(obj1).serializable_hash` caches the generated JSON. Changing `obj1.attr2` currently breaks the whole of the cache for obj1 -- including the included stuff. 
It would be great if it would only break the cache for the specific part of the JSON related to `attr2`.

**Comment by stas** [comment]
>  It would be great if it would only break the cache for the specific part of the JSON related to `attr2`.

@ekampp exactly what would that part be? If a record is updated, any related cache is invalidated.
